servlet/src/main/java/io/grpc/servlet/ForceTrailersServerInterceptor.java

```diff
@@ -0,0 +1,124 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.servlet;
+
+import io.grpc.ForwardingServerCall;
+import io.grpc.ForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+
+class ForceTrailersServerInterceptor implements ServerInterceptor {
+  @Override
+  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call,
+      Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    ForceTrailersServerCall<ReqT, RespT> interceptedCall = new ForceTrailersServerCall<>(call);
+    try {
+      return new ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT>(next.startCall(interceptedCall, headers)) {
+        @Override
+        public void onMessage(ReqT message) {
+          try {
+            super.onMessage(message);
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onHalfClose() {
+          try {
+            super.onHalfClose();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onCancel() {
+          try {
+            super.onCancel();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onComplete() {
+          try {
+            super.onComplete();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onReady() {
+          try {
+            super.onReady();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        private void closeWithThrowable(Throwable t) {
+          Metadata metadata = Status.trailersFromThrowable(t);
+          if (metadata == null) {
+            metadata = new Metadata();
+          }
+          interceptedCall.close(Status.fromThrowable(t), metadata);
+        }
+      };
+    } catch (RuntimeException e) {
+      if (!interceptedCall.headersSent) {
+        call.sendHeaders(new Metadata());
+      }
+      throw e;
+    }
+  }
+
+  static class ForceTrailersServerCall<ReqT, RespT> extends ForwardingServerCall.SimpleForwardingServerCall<ReqT, RespT> {
+
+    boolean headersSent = false;
```

Comment:
Because of the exception handling, which can be happening on a different thread, make this `volatile` and set it to true before calling `super.sendHeaders(headers)`

Reply:
I though of this one and I read the documentation about thread safetyness of the different gRPC components so I was not sure if it needed. Will change this.